File: app/src/main/java/mrkm4ntr/twitterclient/views/DetailFragment.java
package mrkm4ntr.twitterclient.views;

import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.ShareActionProvider;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import mrkm4ntr.twitterclient.R;
import mrkm4ntr.twitterclient.data.TwitterContract;

public class DetailFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {

    private static final String LOG_TAG = DetailFragment.class.getSimpleName();
    public static final String DETAIL_URI = "URI";

    private static final String STATUS_SHARE_HASHTAG = "#TwitterClient";

    private ShareActionProvider mShareActionProvider;
    private String mStatus;
    private Uri mUri;

    private static final int DETAIL_LOADER = 0;

    private static final String[] DETAIL_COLUMNS = {
            TwitterContract.StatusEntry._ID,
            TwitterContract.StatusEntry.COLUMN_USER_PROFILE_IMAGE_URL,
            TwitterContract.StatusEntry.COLUMN_USER_NAME,
            TwitterContract.StatusEntry.COLUMN_USER_SCREEN_NAME,
            TwitterContract.StatusEntry.COLUMN_TEXT,
            TwitterContract.StatusEntry.COLUMN_USER_LOCATION,
            TwitterContract.StatusEntry.COLUMN_USER_BIO
    };

    private ImageView mIconView;
    private TextView mNameView;
    private TextView mScreenNameView;
    private TextView mLocationView;
    private TextView mBioView;
    private TextView mTextView;

    public DetailFragment() {
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Bundle arguments = getArguments();
        if (arguments != null) {
            mUri = arguments.getParcelable(DETAIL_URI);
        }
        View rootView = inflater.inflate(R.layout.fragment_detail, container, false);
        mIconView = (ImageView) rootView.findViewById(R.id.detail_icon);
        mNameView = (TextView) rootView.findViewById(R.id.detail_name_textView);
        mScreenNameView = (TextView) rootView.findViewById(R.id.detail_screenName_textView);
        mLocationView = (TextView) rootView.findViewById(R.id.detail_userLocation_textView);
        mBioView = (TextView) rootView.findViewById(R.id.detail_userBio_textView);
        mTextView = (TextView) rootView.findViewById(R.id.detail_text_textView);
        return rootView;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        getLoaderManager().initLoader(DETAIL_LOADER, null, this);
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (this.isVisible()) {
            inflater.inflate(R.menu.menu_detail_fragment, menu);
            MenuItem menuItem = menu.findItem(R.id.action_share);
            mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(menuItem);
            if (mStatus != null) {
                mShareActionProvider.setShareIntent(createShareStatusIntent());
            }
        }
    }

    private Intent createShareStatusIntent() {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TEXT, mStatus + STATUS_SHARE_HASHTAG);
        return intent;
    }

    @Override
    public Loader onCreateLoader(int id, Bundle args) {
        if (mUri != null) {
            return new CursorLoader(getActivity(), mUri, DETAIL_COLUMNS, null, null, null);
        }
        return null;
    }

    @Override
    public void onLoadFinished(Loader loader, Cursor data) {
        if (data != null && data.moveToFirst()) {
            String profileImageUrl = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_USER_PROFILE_IMAGE_URL));
            new StatusAdapter.UpdateImageViewTask(mIconView, profileImageUrl).execute();
            String name = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_USER_NAME));
            mNameView.setText(name);
            String screenName = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_USER_SCREEN_NAME));
            mScreenNameView.setText(screenName);
            String location = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_USER_LOCATION));
            mLocationView.setText(location);
            String bio = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_USER_BIO));
            mBioView.setText(bio);
            String text = data.getString(data.getColumnIndex(
                    TwitterContract.StatusEntry.COLUMN_TEXT));
            mTextView.setText(text);
            mStatus = text;

            if (mShareActionProvider != null) {
                mShareActionProvider.setShareIntent(createShareStatusIntent());
            }
        }
    }

    @Override
    public void onLoaderReset(Loader loader) {

    }
}

File: app/src/main/kotlin/mrkm4ntr/twitterclient/views/DetailFragment.kt
package mrkm4ntr.twitterclient.views

import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.LoaderManager
import android.support.v4.content.CursorLoader
import android.support.v4.content.Loader
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.ShareActionProvider
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide

import mrkm4ntr.twitterclient.R
import mrkm4ntr.twitterclient.data.TwitterContract
import mrkm4ntr.twitterclient.extensions.applyHttpLink
import mrkm4ntr.twitterclient.extensions.datetimeAgo
import java.util.*

class DetailFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {

    private var mShareActionProvider: ShareActionProvider? = null
    private var mStatus: String? = null
    private var mUri: Uri? = null

    private var mIconView: ImageView? = null
    private var mNameView: TextView? = null
    private var mScreenNameView: TextView? = null
    private var mLocationView: TextView? = null
    private var mBioView: TextView? = null
    private var mTextView: TextView? = null
    private var mCreatedAtView: TextView? = null

    init {
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
            inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val arguments = arguments
        if (arguments != null) {
            mUri = arguments.getParcelable<Uri>(DETAIL_URI)
        }
        val rootView = inflater!!.inflate(R.layout.fragment_detail, container, false)
        mIconView = rootView.findViewById(R.id.detail_icon) as ImageView
        mNameView = rootView.findViewById(R.id.detail_name_textView) as TextView
        mScreenNameView = rootView.findViewById(R.id.detail_screenName_textView) as TextView
        mLocationView = rootView.findViewById(R.id.detail_userLocation_textView) as TextView
        mBioView = rootView.findViewById(R.id.detail_userBio_textView) as TextView
        mTextView = rootView.findViewById(R.id.detail_text_textView) as TextView
        mCreatedAtView = rootView.findViewById(R.id.detail_createAt_textView) as TextView
        return rootView
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        loaderManager.initLoader(DETAIL_LOADER, Bundle.EMPTY, this)
        super.onActivityCreated(savedInstanceState)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        if (this.isVisible) {
            inflater!!.inflate(R.menu.menu_detail_fragment, menu)
            val menuItem = menu!!.findItem(R.id.action_share)
            mShareActionProvider = MenuItemCompat.getActionProvider(menuItem) as ShareActionProvider
            mStatus?.let {
                mShareActionProvider!!.setShareIntent(createShareStatusIntent())
            }
        }
    }

    private fun createShareStatusIntent(): Intent {
        return Intent(Intent.ACTION_SEND).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET)
            setType("text/plain")
            putExtra(Intent.EXTRA_TEXT, mStatus!! + STATUS_SHARE_HASHTAG)
        }
    }

    override fun onCreateLoader(id: Int, args: Bundle): Loader<Cursor>? {
        mUri?.let {
            return CursorLoader(activity, it, DETAIL_COLUMNS, null, null, null)
        }
        return null
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor?) {
        data?.run {
            if (moveToFirst()) {
                val profileImageUrl = getString(getColumnIndex(
                        TwitterContract.StatusEntry.COLUMN_USER_PROFILE_IMAGE_URL))
                Glide.with(this@DetailFragment).load(profileImageUrl).into(mIconView)
                val name = getString(getColumnIndex(TwitterContract.StatusEntry.COLUMN_USER_NAME))
                mNameView!!.text = name
                val screenName = getString(getColumnIndex(
                        TwitterContract.StatusEntry.COLUMN_USER_SCREEN_NAME))
                mScreenNameView!!.text = "@$screenName"
                val location = getString(getColumnIndex(
                        TwitterContract.StatusEntry.COLUMN_USER_LOCATION))
                mLocationView!!.text = location
                val bio = getString(getColumnIndex(TwitterContract.StatusEntry.COLUMN_USER_BIO))
                mBioView!!.text = bio
                val text = getString(getColumnIndex(TwitterContract.StatusEntry.COLUMN_TEXT))
                mTextView!!.text = text
                mTextView!!.applyHttpLink(context)
                mCreatedAtView!!.text =
                        Date(getLong(getColumnIndex(TwitterContract.StatusEntry.COLUMN_CREATE_AT)))
                                .datetimeAgo()

                mStatus = text
                mShareActionProvider?.run {
                    setShareIntent(createShareStatusIntent())
                }
            }
        }
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {

    }

    companion object {

        private val LOG_TAG = DetailFragment::class.java.simpleName
        val DETAIL_URI = "URI"

        private val STATUS_SHARE_HASHTAG = "#TwitterClient"

        private val DETAIL_LOADER = 0

        private val DETAIL_COLUMNS = arrayOf(
                TwitterContract.StatusEntry._ID,
                TwitterContract.StatusEntry.COLUMN_USER_PROFILE_IMAGE_URL,
                TwitterContract.StatusEntry.COLUMN_USER_NAME,
                TwitterContract.StatusEntry.COLUMN_USER_SCREEN_NAME,
                TwitterContract.StatusEntry.COLUMN_TEXT,
                TwitterContract.StatusEntry.COLUMN_USER_LOCATION,
                TwitterContract.StatusEntry.COLUMN_USER_BIO,
                TwitterContract.StatusEntry.COLUMN_CREATE_AT)
    }
}
